File: shared/android/shared/scan/src/main/kotlin/app/releaf/shared/scan/businesscard/pipeline/ExtractionPipeline.kt
/*
 * ExtractionPipeline.kt
 *
 * Orchestrates the extraction flow: classifier fan-out → candidate
 * collection → field resolution → final [ExtractedContact].
 *
 * The pipeline is the only piece that knows the shape of the output
 * — every classifier returns a flat candidate list, and the pipeline
 * decides which candidates win. This keeps the classifier interface
 * simple (single method, single return type) and lets the resolver
 * apply cross-field rules (e.g. "if NAME and DESIGNATION fired on
 * the same block, drop the lower-scoring one").
 *
 * Address is special-cased: AddressClassifier needs the
 * PostcodeClassifier's results, so the pipeline runs postcodes
 * first, then constructs AddressClassifier with that index set.
 */

package app.releaf.shared.scan.businesscard.pipeline

import app.releaf.shared.scan.OcrBlock
import app.releaf.shared.scan.businesscard.ExtractedContact
import app.releaf.shared.scan.businesscard.ExtractionTrace
import app.releaf.shared.scan.businesscard.FieldCandidate
import app.releaf.shared.scan.businesscard.FieldKind
import app.releaf.shared.scan.businesscard.ScoringWeights
import app.releaf.shared.scan.businesscard.classifier.AddressClassifier
import app.releaf.shared.scan.businesscard.classifier.Classifier
import app.releaf.shared.scan.businesscard.classifier.CompanyClassifier
import app.releaf.shared.scan.businesscard.classifier.DesignationClassifier
import app.releaf.shared.scan.businesscard.classifier.EmailClassifier
import app.releaf.shared.scan.businesscard.classifier.NameClassifier
import app.releaf.shared.scan.businesscard.classifier.PhoneClassifier
import app.releaf.shared.scan.businesscard.classifier.PostcodeClassifier
import app.releaf.shared.scan.businesscard.classifier.UrlClassifier

class ExtractionPipeline(
    private val weights: ScoringWeights = ScoringWeights(),
    /**
     * Classifiers run on every pass. Default is the canonical set
     * minus AddressClassifier — that one is constructed inside
     * `extract` because it needs the PostcodeClassifier's output.
     * Callers wanting a different set (tests, benchmarks) can
     * supply their own list.
     */
    private val classifiers: List<Classifier> = listOf(
        EmailClassifier(),
        PhoneClassifier(),
        UrlClassifier(),
        PostcodeClassifier(),
        NameClassifier(),
        DesignationClassifier(),
        CompanyClassifier(),
    ),
) {

    /** Run the pipeline. Set [keepTrace] to populate `result.trace`. */
    fun extract(blocks: List<OcrBlock>, keepTrace: Boolean = false): ExtractedContact {
        if (blocks.isEmpty()) return ExtractedContact.empty
        val timings = mutableMapOf<String, Long>()

        val tLayout = System.nanoTime()
        val layout = layoutOf(blocks)
        timings["layout"] = System.nanoTime() - tLayout

        // Run base classifiers.
        val tClass = System.nanoTime()
        val baseCandidates = classifiers.flatMap { it.classify(layout, weights) }
        timings["classifiers"] = System.nanoTime() - tClass

        // Run AddressClassifier with knowledge of which blocks
        // postcode-classified positively.
        val postcodeBlocks = baseCandidates
            .filter { it.kind == FieldKind.POSTCODE }
            .map { it.sourceBlockIndex }
            .toSet()
        val tAddress = System.nanoTime()
        val addressCandidates = AddressClassifier(postcodeBlocks).classify(layout, weights)
        timings["address"] = System.nanoTime() - tAddress

        val all = baseCandidates + addressCandidates

        // Resolve each field.
        val tResolve = System.nanoTime()
        val resolved = resolve(all, layout)
        timings["resolve"] = System.nanoTime() - tResolve

        return resolved.copy(
            trace = if (keepTrace) ExtractionTrace(all, timings) else null,
        )
    }

    /**
     * Pick the winning candidate(s) per field given the full pool.
     * Single-value fields (NAME / COMPANY / DESIGNATION / ADDRESS)
     * take the top-scoring candidate above the per-kind threshold;
     * multi-value fields (PHONE / EMAIL / WEBSITE) take all
     * candidates above their threshold, deduplicated.
     *
     * When the same block has competing single-value claims (e.g.
     * NAME and COMPANY both fired on the top-of-card text), the
     * higher-scoring kind wins that block and the other kinds skip
     * it. This avoids a name-vs-company battle being decided by
     * the priority order alone.
     */
    private fun resolve(
        candidates: List<FieldCandidate>,
        layout: LayoutContext,
    ): ExtractedContact {
        // Index candidates by kind for the per-kind selectors.
        val byKind = candidates.groupBy { it.kind }

        val emails = pickMulti(
            byKind[FieldKind.EMAIL].orEmpty(),
            min = weights.minEmailScore,
        )
        val phones = pickMulti(
            byKind[FieldKind.PHONE].orEmpty(),
            min = weights.minPhoneScore,
        )
        val websites = pickMulti(
            byKind[FieldKind.WEBSITE].orEmpty(),
            min = weights.minWebsiteScore,
        )

        // Greedy resolution across the three competing single-value
        // kinds (NAME / COMPANY / DESIGNATION). Sort their union by
        // score descending; for each candidate, claim it iff neither
        // the source block nor the kind has been claimed yet.
        // Outcome: a single block can only host one kind, and the
        // strongest cross-kind signal wins each contested block.
        // Address resolves separately because its candidates are
        // cluster-indexed, not single-block.
        val singleValueKinds = setOf(
            FieldKind.NAME, FieldKind.COMPANY, FieldKind.DESIGNATION,
        )
        val minScores = mapOf(
            FieldKind.NAME        to weights.minNameScore,
            FieldKind.COMPANY     to weights.minCompanyScore,
            FieldKind.DESIGNATION to weights.minDesignationScore,
        )
        val claimedBlocks = mutableSetOf<Int>()
        val claimedKinds  = mutableMapOf<FieldKind, FieldCandidate>()
        val singleValueSorted = candidates
            .filter { it.kind in singleValueKinds && it.score >= (minScores[it.kind] ?: Double.MAX_VALUE) }
            .sortedByDescending { it.score }
        for (c in singleValueSorted) {
            if (c.kind in claimedKinds) continue
            if (c.sourceBlockIndex in claimedBlocks) continue
            claimedKinds[c.kind] = c
            claimedBlocks += c.sourceBlockIndex
        }
        val name        = claimedKinds[FieldKind.NAME]
        val company     = claimedKinds[FieldKind.COMPANY]
        val designation = claimedKinds[FieldKind.DESIGNATION]

        val address = pickSingle(
            byKind[FieldKind.ADDRESS].orEmpty(),
            min = weights.minAddressScore,
            avoidBlocks = emptySet(),  // address can overlap — uses cluster index
        )

        // Aggregate confidence: mean of populated fields' scores
        // normalised by an empirical "max plausible score" of 25.
        val populated = listOfNotNull(
            name?.score, company?.score, designation?.score, address?.score,
        ) +
            (byKind[FieldKind.PHONE].orEmpty().filter { it.score >= weights.minPhoneScore }.map { it.score }) +
            (byKind[FieldKind.EMAIL].orEmpty().filter { it.score >= weights.minEmailScore }.map { it.score }) +
            (byKind[FieldKind.WEBSITE].orEmpty().filter { it.score >= weights.minWebsiteScore }.map { it.score })

        val confidence = if (populated.isEmpty()) 0.0
        else (populated.average() / 25.0).coerceIn(0.0, 1.0)

        return ExtractedContact(
            name        = name?.text,
            company     = company?.text,
            designation = designation?.text,
            phones      = phones,
            emails      = emails,
            websites    = websites,
            address     = address?.text,
            confidence  = confidence,
        )
    }

    private fun pickSingle(
        candidates: List<FieldCandidate>,
        min: Double,
        avoidBlocks: Set<Int>,
    ): FieldCandidate? = candidates
        .filter { it.score >= min && it.sourceBlockIndex !in avoidBlocks }
        .maxByOrNull { it.score }

    private fun pickMulti(
        candidates: List<FieldCandidate>,
        min: Double,
    ): List<String> {
        // De-dupe by normalised text, keep first-seen order from the
        // candidates list (already in block order).
        val seen = mutableSetOf<String>()
        val out  = mutableListOf<String>()
        for (c in candidates) {
            if (c.score < min) continue
            val key = c.text.lowercase()
            if (seen.add(key)) out += c.text
        }
        return out
    }
}

File: shared/ios/ReleafCore/Sources/ReleafCoreScan/BusinessCard/Pipeline/ExtractionPipeline.swift
/*
 * ExtractionPipeline.swift  — mirror of ExtractionPipeline.kt.
 */

import Foundation

public struct ExtractionPipeline: Sendable {

    private let weights: ScoringWeights
    private let classifiers: [Classifier]

    public init(
        weights: ScoringWeights = ScoringWeights(),
        classifiers: [Classifier] = [
            EmailClassifier(),
            PhoneClassifier(),
            UrlClassifier(),
            PostcodeClassifier(),
            NameClassifier(),
            DesignationClassifier(),
            CompanyClassifier(),
        ]
    ) {
        self.weights     = weights
        self.classifiers = classifiers
    }

    public func extract(_ blocks: [OcrBlock], keepTrace: Bool = false) -> ExtractedContact {
        guard !blocks.isEmpty else { return .empty }
        var timings: [String: UInt64] = [:]

        // ML Kit emits both Paragraph- and Line-grained blocks; the
        // Paragraph block concatenates every line on the card and
        // poisons the classifier scores. Filter to Line when present.
        // Vision only emits Line, so this is a no-op there. Mirror
        // of the Android pipeline.
        let preferred: [OcrBlock] = {
            if blocks.contains(where: { $0.kind == .line }) {
                return blocks.filter { $0.kind == .line }
            }
            return blocks
        }()
        guard !preferred.isEmpty else { return .empty }

        let t0 = DispatchTime.now().uptimeNanoseconds
        let layout = layoutOf(preferred)
        timings["layout"] = DispatchTime.now().uptimeNanoseconds - t0

        let t1 = DispatchTime.now().uptimeNanoseconds
        var baseCandidates: [FieldCandidate] = []
        for c in classifiers {
            baseCandidates.append(contentsOf: c.classify(layout: layout, weights: weights))
        }
        timings["classifiers"] = DispatchTime.now().uptimeNanoseconds - t1

        // Address — needs postcode results.
        let postcodeBlocks = Set(
            baseCandidates.filter { $0.kind == .postcode }.map(\.sourceBlockIndex)
        )
        let t2 = DispatchTime.now().uptimeNanoseconds
        let addressCandidates = AddressClassifier(postcodeBlockIndices: postcodeBlocks)
            .classify(layout: layout, weights: weights)
        timings["address"] = DispatchTime.now().uptimeNanoseconds - t2

        let allCandidates = baseCandidates + addressCandidates

        // Drop candidates whose own text reads as the category
        // name (or one of its OCR-error aliases). Mirror of the
        // Android pipeline filter — same stop words, same
        // alphanumeric-only normalisation.
        let filteredCandidates = allCandidates.filter { candidate in
            !Self.categoryStopWords.contains(Self.normaliseForStopCheck(candidate.text))
        }

        // Cross-kind layout bonus — name-above-designation pattern.
        // Mirror of the Android pipeline.
        let boostedCandidates = applyNameDesignationAdjacencyBonus(
            filteredCandidates,
            layout: layout
        )

        let t3 = DispatchTime.now().uptimeNanoseconds
        let resolved = resolve(boostedCandidates)
        timings["resolve"] = DispatchTime.now().uptimeNanoseconds - t3

        let trace = keepTrace
            ? ExtractionTrace(candidates: allCandidates, timings: timings)
            : nil

        return ExtractedContact(
            name:        resolved.name,
            company:     resolved.company,
            designation: resolved.designation,
            phones:      resolved.phones,
            emails:      resolved.emails,
            websites:    resolved.websites,
            address:     resolved.address,
            confidence:  resolved.confidence,
            trace:       trace
        )
    }

    private func resolve(_ candidates: [FieldCandidate]) -> ExtractedContact {
        let byKind: [FieldKind: [FieldCandidate]] = Dictionary(grouping: candidates) { $0.kind }

        let emails  = pickMulti(byKind[.email]   ?? [], min: weights.minEmailScore)
        let phones  = pickMulti(byKind[.phone]   ?? [], min: weights.minPhoneScore)
        let websites = pickMulti(byKind[.website] ?? [], min: weights.minWebsiteScore)

        // Greedy cross-kind resolution. Sort the union of single-value
        // candidates by score desc; claim each iff neither the source
        // block nor the kind has been claimed yet. Mirror of Android's
        // resolver logic.
        let singleValueKinds: Set<FieldKind> = [.name, .company, .designation]
        let mins: [FieldKind: Double] = [
            .name:        weights.minNameScore,
            .company:     weights.minCompanyScore,
            .designation: weights.minDesignationScore,
        ]
        var claimedBlocks: Set<Int> = []
        var claimedKinds: [FieldKind: FieldCandidate] = [:]
        let pool = candidates
            .filter { singleValueKinds.contains($0.kind) && $0.score >= (mins[$0.kind] ?? .greatestFiniteMagnitude) }
            .sorted { $0.score > $1.score }
        for c in pool {
            if claimedKinds[c.kind] != nil { continue }
            if claimedBlocks.contains(c.sourceBlockIndex) { continue }
            claimedKinds[c.kind] = c
            claimedBlocks.insert(c.sourceBlockIndex)
        }
        let name        = claimedKinds[.name]
        let company     = claimedKinds[.company]
        let designation = claimedKinds[.designation]

        let address = pickSingle(byKind[.address] ?? [], min: weights.minAddressScore, avoid: [])

        var populated: [Double] = []
        if let n = name        { populated.append(n.score) }
        if let c = company     { populated.append(c.score) }
        if let d = designation { populated.append(d.score) }
        if let a = address     { populated.append(a.score) }
        populated.append(contentsOf: (byKind[.phone] ?? [])
            .filter { $0.score >= weights.minPhoneScore }.map(\.score))
        populated.append(contentsOf: (byKind[.email] ?? [])
            .filter { $0.score >= weights.minEmailScore }.map(\.score))
        populated.append(contentsOf: (byKind[.website] ?? [])
            .filter { $0.score >= weights.minWebsiteScore }.map(\.score))

        let confidence: Double
        if populated.isEmpty {
            confidence = 0
        } else {
            let mean = populated.reduce(0, +) / Double(populated.count)
            confidence = max(0, min(1, mean / 25.0))
        }

        return ExtractedContact(
            name:        name?.text,
            company:     company?.text,
            designation: designation?.text,
            phones:      phones,
            emails:      emails,
            websites:    websites,
            address:     address?.text,
            confidence:  confidence
        )
    }

    private func pickSingle(
        _ candidates: [FieldCandidate],
        min: Double,
        avoid: Set<Int>
    ) -> FieldCandidate? {
        candidates
            .filter { $0.score >= min && !avoid.contains($0.sourceBlockIndex) }
            .max(by: { $0.score < $1.score })
    }

    private func pickMulti(
        _ candidates: [FieldCandidate],
        min: Double
    ) -> [String] {
        var seen: Set<String> = []
        var out: [String] = []
        for c in candidates {
            if c.score < min { continue }
            let key = c.text.lowercased()
            if seen.insert(key).inserted { out.append(c.text) }
        }
        return out
    }

    /// Boost NAME / DESIGNATION pairs where the name's bbox sits
    /// directly above the designation's bbox. The pattern is the
    /// single most reliable layout cue on business cards. Adds
    /// `nameDesignationAdjacencyBonus` to BOTH candidates' scores
    /// when matched. Mirror of `applyNameDesignationAdjacencyBonus`
    /// in the Android pipeline.
    private func applyNameDesignationAdjacencyBonus(
        _ candidates: [FieldCandidate],
        layout: LayoutContext
    ) -> [FieldCandidate] {
        let names  = candidates.filter { $0.kind == .name }
        let desigs = candidates.filter { $0.kind == .designation }
        if names.isEmpty || desigs.isEmpty { return candidates }

        var nameBoosts:  Set<Int> = []
        var desigBoosts: Set<Int> = []
        for n in names {
            let nBox = layout.blocks[n.sourceBlockIndex].bbox
            let nBottom = nBox.y + nBox.height
            for d in desigs {
                let dBox = layout.blocks[d.sourceBlockIndex].bbox
                let gap = dBox.y - nBottom
                if gap >= 0, gap <= weights.adjacencyYDistance {
                    nameBoosts.insert(n.sourceBlockIndex)
                    desigBoosts.insert(d.sourceBlockIndex)
                }
            }
        }

        if nameBoosts.isEmpty { return candidates }
        return candidates.map { c in
            if c.kind == .name && nameBoosts.contains(c.sourceBlockIndex) {
                return FieldCandidate(
                    sourceBlockIndex: c.sourceBlockIndex,
                    text:             c.text,
                    kind:             c.kind,
                    score:            c.score + weights.nameDesignationAdjacencyBonus
                )
            }
            if c.kind == .designation && desigBoosts.contains(c.sourceBlockIndex) {
                return FieldCandidate(
                    sourceBlockIndex: c.sourceBlockIndex,
                    text:             c.text,
                    kind:             c.kind,
                    score:            c.score + weights.nameDesignationAdjacencyBonus
                )
            }
            return c
        }
    }

    /// Strings that should never surface as an extracted field on a
    /// Business Card capture. Compared against each candidate's
    /// alphanumeric-only lower-cased text so "Business Card",
    /// "Business-Card", "businesscard.", and the "8usiness" OCR-
    /// error aliases all collapse to the same key.
    private static let categoryStopWords: Set<String> = [
        "businesscard",
        "card",
        "business",
        "8usinesscard",
        "8usiness",
    ]

    private static func normaliseForStopCheck(_ s: String) -> String {
        s.lowercased().filter { $0.isLetter || $0.isNumber }
    }
}
